File: android/auto-mobile-sdk/src/main/kotlin/dev/jasonpearson/automobile/sdk/events/SdkEventBuffer.kt
package dev.jasonpearson.automobile.sdk.events

import dev.jasonpearson.automobile.protocol.SdkEvent
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Thread-safe buffer for SDK events that flushes on capacity or timer.
 *
 * Events are collected and flushed as a batch to reduce Intent broadcast frequency.
 * Flush occurs when [maxBufferSize] is reached or every [flushIntervalMs] milliseconds,
 * whichever comes first.
 *
 * @param maxBufferSize Maximum events before forced flush (default 50)
 * @param flushIntervalMs Periodic flush interval in milliseconds (default 500)
 * @param onFlush Callback invoked with the batch of events to send
 * @param executor Optional executor for periodic flush scheduling (for testing)
 */
class SdkEventBuffer(
  private val maxBufferSize: Int = 50,
  private val flushIntervalMs: Long = 500,
  private val onFlush: (List<SdkEvent>) -> Unit,
  private val executor: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor { r ->
    Thread(r, "SdkEventBuffer").apply { isDaemon = true }
  },
) {
  private val lock = ReentrantLock()
  private val buffer = mutableListOf<SdkEvent>()
  private var flushTask: ScheduledFuture<*>? = null
  @Volatile private var isShutdown = false

  /** Start the periodic flush timer. */
  fun start() {
    lock.withLock {
      if (flushTask == null && !isShutdown) {
        flushTask = executor.scheduleAtFixedRate(
          { flush() },
          flushIntervalMs,
          flushIntervalMs,
          TimeUnit.MILLISECONDS,
        )
      }
    }
  }

  /** Add an event to the buffer. Flushes immediately if buffer is full. */
  fun add(event: SdkEvent) {
    if (isShutdown) return

    val shouldFlush: Boolean
    val snapshot: List<SdkEvent>

    lock.withLock {
      buffer.add(event)
      if (buffer.size >= maxBufferSize) {
        snapshot = ArrayList(buffer)
        buffer.clear()
        shouldFlush = true
      } else {
        snapshot = emptyList()
        shouldFlush = false
      }
    }

    if (shouldFlush) {
      deliverBatch(snapshot)
    }
  }

  /** Flush any buffered events immediately. */
  fun flush() {
    val snapshot: List<SdkEvent>

    lock.withLock {
      if (buffer.isEmpty()) return
      snapshot = ArrayList(buffer)
      buffer.clear()
    }

    deliverBatch(snapshot)
  }

  /** Shutdown the buffer, flushing remaining events. */
  fun shutdown() {
    isShutdown = true
    flushTask?.cancel(false)
    flush()
    executor.shutdown()
  }

  private fun deliverBatch(events: List<SdkEvent>) {
    if (events.isEmpty()) return
    try {
      onFlush(events)
    } catch (_: Exception) {
      // Swallow exceptions to prevent caller disruption
    }
  }
}

File: ios/auto-mobile-sdk/Sources/AutoMobileSDK/Events/SdkEventBuffer.swift
import Foundation

/// Protocol for event buffering to allow faking in tests.
public protocol EventBuffering: AnyObject, Sendable {
    func add(_ event: any SdkEvent)
    func start()
    func shutdown()
    func flush()
}

/// Thread-safe event buffer that flushes on capacity or timer.
public final class SdkEventBuffer: EventBuffering, @unchecked Sendable {
    private let maxBufferSize: Int
    private let flushIntervalMs: Int
    private let onFlush: @Sendable ([any SdkEvent]) -> Void
    private let lock = NSLock()
    private var buffer: [any SdkEvent] = []
    private var timer: (any TimerScheduling)?
    private let timerFactory: () -> any TimerScheduling

    public init(
        maxBufferSize: Int = 50,
        flushIntervalMs: Int = 500,
        timerFactory: @escaping () -> any TimerScheduling = { GCDTimer() },
        onFlush: @escaping @Sendable ([any SdkEvent]) -> Void
    ) {
        self.maxBufferSize = maxBufferSize
        self.flushIntervalMs = flushIntervalMs
        self.timerFactory = timerFactory
        self.onFlush = onFlush
    }

    public func start() {
        lock.lock()
        defer { lock.unlock() }
        guard timer == nil else { return }
        let t = timerFactory()
        timer = t
        t.schedule(intervalMs: flushIntervalMs) { [weak self] in
            self?.flush()
        }
    }

    public func add(_ event: any SdkEvent) {
        var shouldFlush = false
        lock.lock()
        buffer.append(event)
        shouldFlush = buffer.count >= maxBufferSize
        lock.unlock()
        if shouldFlush {
            flush()
        }
    }

    public func flush() {
        lock.lock()
        guard !buffer.isEmpty else {
            lock.unlock()
            return
        }
        let events = buffer
        buffer.removeAll(keepingCapacity: true)
        lock.unlock()
        onFlush(events)
    }

    public func shutdown() {
        lock.lock()
        timer?.cancel()
        timer = nil
        let remaining = buffer
        buffer.removeAll()
        lock.unlock()
        if !remaining.isEmpty {
            onFlush(remaining)
        }
    }
}

// MARK: - Timer Abstraction

/// Protocol for timer scheduling to allow faking in tests.
public protocol TimerScheduling: AnyObject, Sendable {
    func schedule(intervalMs: Int, block: @escaping @Sendable () -> Void)
    func cancel()
}

/// GCD-based timer implementation.
public final class GCDTimer: TimerScheduling, @unchecked Sendable {
    private var source: DispatchSourceTimer?
    private let queue = DispatchQueue(label: "dev.jasonpearson.automobile.sdk.timer")

    public init() {}

    public func schedule(intervalMs: Int, block: @escaping @Sendable () -> Void) {
        let source = DispatchSource.makeTimerSource(queue: queue)
        source.schedule(
            deadline: .now() + .milliseconds(intervalMs),
            repeating: .milliseconds(intervalMs)
        )
        source.setEventHandler(handler: block)
        source.resume()
        self.source = source
    }

    public func cancel() {
        source?.cancel()
        source = nil
    }

    deinit {
        source?.cancel()
    }
}
